### PR TITLE
region_scatter: fix a bug that the leaders may be unbalanced after scatter region

### LINKED ISSUE #6017: tikv leader are unbalanced during br full restore data
### What did you do?
6 tikvs（8c/32g）with az label
br full restore 20000 warehouse
https://tcms.pingcap.net/dashboard/executions/plan/1562736

### What did you expect to see?
pd leader are balanced during br full restore data 
8da2338a-947a-456c-979f-6a91dcc8569b

### What did you see instead?
pd leader are unbalanced during br full restore data 

### What version of PD are you using (`pd-server -V`)?
./pd-server -V
 Release Version: v6.6.0
Edition: Community
Git Commit Hash: 67c0fa283673257b4dc0839f19cd74fec5ef46fb
Git Branch: heads/refs/tags/v6.6.0
UTC Build Time:  2023-02-14 11:56:24
2023-02-17T03:25:41.343Z	INFO	k8s/client.go:132

### PR DESCRIPTION
**Pull request by HunDunDM**
### What problem does this PR solve?

Issue Number: Close #6017 

### What is changed and how does it work?

Use a more correct method to select candidate leaders.

### Check List

Tests

- Unit test

Code changes

Side effects

Related changes

- Need to cherry-pick to the release branch

### Release note

```release-note
Fix a bug that the leaders may be unbalanced during scatter region.
```

### CODE REVIEW

File: server/schedule/region_scatterer.go [hunk 1]
```diff
@@ -322,15 +323,19 @@ func (r *RegionScatterer) scatterRegion(region *core.RegionInfo, group string) *
 	}
 
 	targetPeers := make(map[uint64]*metapb.Peer, len(region.GetPeers()))                  // StoreID -> Peer
-	selectedStores := make(map[uint64]struct{}, len(region.GetPeers()))                   // StoreID set
+	selectedStores := make(map[uint64]struct{}, len(region.GetPeers()))                   // selected StoreID set
+	leaderCandidateStores := make([]uint64, 0, len(region.GetPeers()))                    // StoreID allowed to become Leader
 	scatterWithSameEngine := func(peers map[uint64]*metapb.Peer, context engineContext) { // peers: StoreID -> Peer
 		for _, peer := range peers {
 			if _, ok := selectedStores[peer.GetStoreId()]; ok {
+				if allowLeader(oldFit, peer) {
```
Comment (CabinfeverB) [inline server/schedule/region_scatterer.go:331]: Maybe we can use func `allowLeader` to check `targetPeers`. It seems more straightforward

**Review by CabinfeverB — APPROVED** [review]
Rest LGTM

**Review by lhy1024 — APPROVED** [review]

File: server/schedule/region_scatterer.go (hunk 1 above)
Comment (HunDunDM) [reply server/schedule/region_scatterer.go:331]: The parameter of the `allowLeader` function needs to be a source peer.

**Pull request merged**